python/ray/dag/compiled_dag_node.py

```diff
@@ -1681,52 +1822,24 @@ def _generate_dag_operation_graph_node(
                 dag_node = self.idx_to_task[task_idx].dag_node
                 method_name = exec_task.method_name
                 actor_handle = dag_node._get_actor_handle()
-                requires_nccl = dag_node.type_hint.requires_nccl()
-                upstream_requires_nccl = False
-                for upstream_node in dag_node._upstream_nodes:
-                    if upstream_node.type_hint.requires_nccl():
-                        upstream_requires_nccl = True
-                        break
-
-                read_node = _DAGOperationGraphNode(
-                    _DAGNodeOperation(
-                        exec_task_idx, _DAGNodeOperationType.READ, method_name
-                    ),
-                    task_idx,
-                    actor_handle,
-                    upstream_requires_nccl,
-                )
+                # [TODO] Set all three requires_nccl_*.
+                requires_nccl_read = dag_node.requires_nccl_read
+                requires_nccl_write = dag_node.requires_nccl_write
+                requires_nccl_collective = dag_node.requires_nccl_collective
```

Comment:
Why do we need three different flags instead of just `requires_nccl`?

Reply:
The current implementation of overlapping GPU communication supports overlapping NCCL read with computation. The NCCL collectives (only allreduce so far) are currently synchronized.

To overlap NCCL read with computation, during scheduling, a NCCL read node is swapped with its previous compute node if possible. That is a case where we need `requires_nccl_read` or some similar sort of bookkeeping.

We can probably combine these three different flags into one `requires_nccl` if we also support overlapping NCCL collectives with computation. Should we implement that in this same PR?

Reply:
Yes, let's do that in this PR. So the new rule should be that `requires_nccl` nodes cannot be swapped with each other but can be swapped with other nodes. Does that sound right?